### PR TITLE
Fix missing value in password fields

### PR DESCRIPTION
**Pull request by cconard96**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | Ref https://forum.glpi-project.org/viewtopic.php?id=285362

Password fields from the form field twig template were not having their value set. This broke the display of the simcard PIN/PUK fields for example.

### CODE REVIEW

**Review by orthagh — APPROVED** [review]

**Comment by trasher** [comment]
I wonder if setting this value will not cause issues with some real password fields.

**Review by cedric-anne — CHANGES_REQUESTED** [review]
Adding the existing value in DOM may is not correct from a security point of view.

The solution for this problem is probably to use same kind of logic as we already do for some password fields, like the `passwd` field of `MailCollector`, i.e., adding a checkbox to ask if password should be cleared, and redefine the value only if password field is not empty.

**Comment by cconard96** [comment]
> Adding the existing value in DOM may is not correct from a security point of view.
> 
> The solution for this problem is probably to use same kind of logic as we already do for some password fields, like the `passwd` field of `MailCollector`, i.e., adding a checkbox to ask if password should be cleared, and redefine the value only if password field is not empty.

These fields shouldn't be treated like passwords at all then.
Another type of field should be used for secrets that need to be seen and not write-only.

**Comment by cedric-anne** [comment]
> > Adding the existing value in DOM may is not correct from a security point of view.
> > The solution for this problem is probably to use same kind of logic as we already do for some password fields, like the `passwd` field of `MailCollector`, i.e., adding a checkbox to ask if password should be cleared, and redefine the value only if password field is not empty.
> 
> These fields shouldn't be treated like passwords at all then. Another type of field should be used for secrets that need to be seen and not write-only.

So, in this case, we should probably use a `<input type="text">` field.

**Comment by cconard96** [comment]
Alternatively, any password/secret that doesn't intend to have its value shown should just not pass the value to that password field macro.

**Comment by trasher** [comment]
> Alternatively, any password/secret that doesn't intend to have its value shown should just not pass the value to that password field macro.

Yes, but it's difficult to be sure it's the case. I thunk using another type of field for non password is the best way to go (no surprise).

**Comment by cedric-anne** [comment]
It is probable that same issue is present in all usage of `passwordField` macro.

To handle it in a secure way, I think we should handle the `disclose` feature directly in the `passwordField` macro, by adding a `is_disclosable` boolean option (default to `false`). The `value` attribute would be defined only for disclosable fields, so we would not expose this value in DOM if user is not supposed to be able to read it.

To handle non disclosable fields, we should probably add a `clearable` option (default to `!is_disclosable`), that will add a `<input type="checkbox" name="_blank_{{ name }}" />`, and check all usages of this macro to ensure that password is cleared only when this field is set to 1.

**Review by trasher — APPROVED** [review]

File: templates/components/form/fields_macros.html.twig [hunk 1]
```diff
@@ -400,15 +400,41 @@
...
          {{ options.required ? 'required' : '' }} />
    {% endset %}
+   {% set btn_group %}
+      &nbsp;<i class="far fa-eye pointer disclose" onmousedown="showField('{{ options.id }}')"
+               onmouseup="hideField('{{ options.id }}')"
+               onmouseout="hideField('{{ options.id }}')"></i>
+      &nbsp;<i class="fa fa-paste pointer disclose" onclick="copyToClipboard('{{ options.id }}')"></i>
+   {% endset %}
+   {% if options.is_disclosable %}
+      {%  set label = label ~ btn_group %}
```
Comment (cedric-anne) [inline templates/components/form/fields_macros.html.twig:429]:
```suggestion
      {% set label = label ~ btn_group %}
```

File: templates/components/form/fields_macros.html.twig [hunk 2]
```diff
@@ -400,15 +400,41 @@
...
+             value="{{ options.is_disclosable ? value : '' }}"
          {{ options.readonly ? 'readonly' : '' }}
          {{ options.disabled ? 'disabled' : '' }}
          {{ options.required ? 'required' : '' }} />
    {% endset %}
+   {% set btn_group %}
+      &nbsp;<i class="far fa-eye pointer disclose" onmousedown="showField('{{ options.id }}')"
+               onmouseup="hideField('{{ options.id }}')"
+               onmouseout="hideField('{{ options.id }}')"></i>
+      &nbsp;<i class="fa fa-paste pointer disclose" onclick="copyToClipboard('{{ options.id }}')"></i>
```
Comment (cedric-anne) [inline templates/components/form/fields_macros.html.twig:426]: `showField`, `hideField` and `copyToClipboard` are declared in `templates/components/form/item_device.html.twig` and will not be accessible in some contexts. The should be moved in `common.js` file.

**Review by cedric-anne — CHANGES_REQUESTED** [review]

**Review by cedric-anne — APPROVED** [review]

**Comment by efgiese** [comment]
Can you please check the patch again?
With my installation (upgrade from 9.5.6 to 10.0.5), the points are now displayed, but the display or copying of the PIN/PUK does not work.

**Comment by cedric-anne** [comment]
@efgiese This patch will be part of the 10.0.6 version that should be released soon.

**Comment by efgiese** [comment]
I've applied the patch manually but it doesn't seem to work entirely.

**Comment by trasher** [comment]
Use provided nightly builds

**Pull request merged**